File: chap01_biochem/QSSA_enzyme.m
function enzyme
% asymptotic analysis of enzyme kinetics equations - QSSA
clear all
close all
clc
set(0,                           ...
   'defaultaxesfontsize', 20,   ...
   'defaultaxeslinewidth', 2.0, ...
   'defaultlinelinewidth', 2.0);
   
% Part 2:  QSS approximation
par.a = 0.5;
par.k = 1.5;
par.e = 0.05;

tspan = linspace(0,20,2000);
initial = [1,0];
[t,Y] = ode15s(@(t,y)rhsqss(t,y,par),tspan,initial);
s=Y(:,1);
x=Y(:,2);
 
figure(4)
plot(t,s,t,x,'linewidth',2)
legend('\sigma','z','fontsize',18)
xlabel('\eta','fontsize',20)
axis([0 4 0 1])
% now plot the slow manifold and the solution together
slow_s = linspace(0,1.2,100);
slow_x = slow_s./(slow_s+par.k) ;
figure(5)
plot(s,x,slow_s,slow_x,'--','linewidth',2)
xlabel('\sigma','fontsize',20)
ylabel('x','fontsize',20)
axis([0 1.2 0 0.5])
 
forplotting3 = [t s x];
save('temp3.dat','forplotting3')
forplotting4 = [slow_s ;  slow_x]';
save('temp4.dat','forplotting4')


end

 

function out = rhsqss(t,y,par)
s = y(1);
x = y(2);
out(1) = -s+x*(s+par.a) ;
out(2) = (s-x*(s+par.k))/par.e;
out = out';  % don't forget this or you get rubbish.
end

File: chap01_biochem/QSSA_enzyme.py
#-------------------------------------------------------------------

# Python code for simulating the quasi-steady-state approximation (QSSA) 
# of enzyme kinetics.

# For Chapter 1 of
# Keener and Sneyd, Mathematical Physiology, 3rd Edition, Springer.

# Written by James Keener and James Sneyd

#-------------------------------------------------------------------

import numpy as np
import matplotlib.pyplot as plt
from scipy.integrate import odeint

def rhsqss(y, t, a, k, e):
    s, x = y
    dsdt = -s + x * (s + a)
    dxdt = (s - x * (s + k)) / e
    return [dsdt, dxdt]



# Part 2: QSS approximation
a = 0.5
k = 1.5
e = 0.05

tspan = np.linspace(0, 20, 2000)
initial = [1, 0]
sol = odeint(rhsqss, initial, tspan, args=(a, k, e))

s = sol[:, 0]
x = sol[:, 1]

# Plotting
plt.figure(4)
plt.plot(tspan, s, tspan, x, linewidth=2)
plt.legend(['$\sigma$', 'x'], fontsize=18)
plt.xlabel('$\eta$', fontsize=20)
plt.axis([0, 4, 0, 1])

# Plot the slow manifold and the solution together
slow_s = np.linspace(0, 1.2, 100)
slow_x = slow_s / (slow_s + k)
plt.figure(5)
plt.plot(s, x, slow_s, slow_x, '--', linewidth=2)
plt.xlabel('$\sigma$', fontsize=20)
plt.ylabel('x', fontsize=20)
plt.axis([0, 1.2, 0, 0.5])

np.savetxt('temp3.dat', np.column_stack((tspan, s, x)), header='t s x')
np.savetxt('temp4.dat', np.column_stack((slow_s, slow_x)), header='s x')

plt.show()
